### PR TITLE
Multi agent support

### PR DESCRIPTION
**Pull request by braelyn-ai**
## 📥 Pull Request

- Creates Agent decorator
  - Can be used to assign classes or functions as agents
- Using the decorator creates an agent in the db
- Singleton decorator
- Makes `Client` adhere to a singleton pattern
- Jupyter Notebook example

Small changes:
- Removes frequency recording in env as it does not work on M1 chip macs
- `warn` -> `warning`
- `print` -> `logging`

### CODE REVIEW

File: agentops/llm_tracker.py [hunk 1]
```diff
@@ -3,10 +3,27 @@
...
+        # Get the locals from the current frame
+        local_vars = frame_info.frame.f_locals
+        for var in local_vars.values():
+            if var == "__main__":
+                return
+            # Instead of checking if var is a class, check if it's an instance with the _is_ao_agent attribute
+            if hasattr(var, '_is_ao_agent') and getattr(var, '_is_ao_agent'):
+                logging.info('LLM call from agent named: ' + getattr(var, '_ao_agent_name'))
+                return getattr(var, '_ao_agent_id')
+    return None
```
Comment (siyangqiu) [inline agentops/llm_tracker.py]: The comments are outdated <-- which is why I typically don't like them. But in this case, I think we should leave an explanation as to what this whole thing is for and another comment explaining that the `var == __main__` piece.

File: agentops/llm_tracker.py [hunk 2]
```diff
@@ -59,7 +78,7 @@ def handle_stream_chunk(chunk):
                     self.client.record(self.event_stream)
                     self.event_stream = None
             except Exception as e:
-                print(
+                logging.info(
```
Comment (siyangqiu) [inline agentops/llm_tracker.py]: I have all of the my logs start with "AgentOps:". I think we should try to be consistent. Also, I think this deserves at least a logging.warning().

File: agentops/llm_tracker.py [hunk 3]
```diff
@@ -115,7 +135,7 @@ def generator():
                 ))
                 # Standard response
             except Exception as e:
-                print(
+                logging.error(
                     f"Unable to parse a chunk for LLM call {kwargs} - skipping upload to AgentOps")
```
Comment (siyangqiu) [inline agentops/llm_tracker.py]: same as above, except the logging level isn't consistent

File: agentops/llm_tracker.py [hunk 4]
```diff
@@ -165,7 +187,7 @@ def handle_stream_chunk(chunk: ChatCompletionChunk):
                     self.client.record(self.event_stream)
                     self.event_stream = None
             except Exception as e:
-                print(
+                logging.error(
                     f"Unable to parse a chunk for LLM call {kwargs} - skipping upload to AgentOps")
```
Comment (siyangqiu) [inline agentops/llm_tracker.py]: same

File: agentops/llm_tracker.py [hunk 5]
```diff
@@ -210,7 +234,7 @@ async def async_generator():
             ))
             # Standard response
         except Exception as e:
-            print(
+            logging.error(
                 f"Unable to parse a chunk for LLM call {kwargs} - skipping upload to AgentOps")
```
Comment (siyangqiu) [inline agentops/llm_tracker.py:221]: same

File: examples/multi_agent_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+     "text": [
+      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
+      "INFO:root:LLM call from agent named: engineer\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Unable to parse a chunk for LLM call {'model': 'gpt-3.5-turbo', 'messages': [{'role': 'system', 'content': 'You are a software engineer and only output python code, no markdown tags.'}, {'role': 'user', 'content': 'Write a python function that accepts two numbers and multiplies them together, then divides by two. No example.'}], 'temperature': 0.5} - skipping upload to AgentOps\n"
```
Comment (siyangqiu) [inline examples/multi_agent_example.ipynb]: You should probably clear the outputs and save?

File: pyproject.toml [hunk 7]
```diff
@@ -7,7 +7,9 @@ name = "agentops"
 version = "0.0.20"
 authors = [
   { name="Alex Reibman", email="areibman@gmail.com" },
-  { name="Shawn Qiu", email="siyangqiu@gmail.com" }
+  { name="Shawn Qiu", email="siyangqiu@gmail.com" },
+  { name="Braelyn Boynton", email="bboynton97@gmail.com" },
+  { name="Howard Gil", email="howardbgil@gmail.com" }
```
Comment (siyangqiu) [inline pyproject.toml:12]: NICE!!!

File: agentops/llm_tracker.py [hunk 8]
```diff
@@ -3,10 +3,27 @@
...
+    # whatever agent class was used last
+    for frame_info in inspect.stack():
+        # Get the locals from the current frame
+        local_vars = frame_info.frame.f_locals
+        for var in local_vars.values():
+            if var == "__main__":
+                return
+            # Instead of checking if var is a class, check if it's an instance with the _is_ao_agent attribute
+            if hasattr(var, '_is_ao_agent') and getattr(var, '_is_ao_agent'):
+                logging.info('LLM call from agent named: ' + getattr(var, '_ao_agent_name'))
```
Comment (siyangqiu) [inline agentops/llm_tracker.py]: Do we want this logging to show up? Maybe logging.debug() so it's not on by default?

**Review by siyangqiu — APPROVED** [review]
Some nits, but 🥳🥳🥳🥳

File: agentops/llm_tracker.py (hunk 1 above)
Comment (braelyn-ai) [reply agentops/llm_tracker.py]: probably smart

File: agentops/llm_tracker.py (hunk 2 above)
Comment (braelyn-ai) [reply agentops/llm_tracker.py]: bet

Comment (braelyn-ai) [reply agentops/llm_tracker.py]: I didn't write the log, just switched it from print, but I can add it

File: examples/multi_agent_example.ipynb (hunk 6 above)
Comment (braelyn-ai) [reply examples/multi_agent_example.ipynb]: My thought was to leave outputs in so you can see the flow without running it. If we don't want to do that, I can clear it

Comment (braelyn-ai) [reply examples/multi_agent_example.ipynb]: i cleared it and added more explanation in the text

**Comment by braelyn-ai** [comment]
- Tracks events to agent_id now
- Changes `_ao_...` to `_agent_ops_...`
- Uses `agent_id` instead of `is_agent`

**Pull request merged**